### PR TITLE
fix: [PL-39219]: Add more shuffling to delegate selection

### PR DESCRIPTION
**Pull request by jennyjames21**
## Describe your changes
Summary:
Add more shuffling for better random results while doing fair distribution of task among delegates. 

Test
- PR testing with parallel step executions and checked the distribution among delegates

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48901)

### CODE REVIEW

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger messagecheck

**Comment by jennyjames21** [comment]
trigger messagecheck

File: 400-rest/src/test/java/software/wings/service/impl/DelegateTaskServiceClassicImplTest.java [hunk 1]
```diff
@@ -72,6 +74,36 @@ public void sortedEligibleListTest() throws ExecutionException {
...
 
+  @Test
+  @Owner(developers = JENNY)
+  @Category(UnitTests.class)
+  public void sortedEligibleListTestRand() throws ExecutionException {
+    String accountId = generateUuid();
+    Delegate delegate1 = createDelegate(accountId, "delelgate1");
+    Delegate delegate2 = createDelegate(accountId, "delegate2");
+    Delegate delegate3 = createDelegate(accountId, "delegate3");
+    Delegate delegate4 = createDelegate(accountId, "delegate4");
```
Comment (gn-harness) [inline 400-rest/src/test/java/software/wings/service/impl/DelegateTaskServiceClassicImplTest.java:85]: Please increase the number of delegates to deflake the test

File: 400-rest/src/main/java/software/wings/service/impl/DelegateTaskServiceClassicImpl.java [hunk 2]
```diff
@@ -290,12 +291,12 @@ public class DelegateTaskServiceClassicImpl implements DelegateTaskServiceClassi
...
-  final Comparator<Map.Entry<String, Long>> delegateTaskCountComparator = (d1, d2) -> {
-    long diff = d1.getValue() - d2.getValue();
+  final Comparator<Map.Entry<String, Integer>> delegateTaskCountComparator = (d1, d2) -> {
+    Integer diff = d1.getValue() - d2.getValue();
     if (diff == 0) {
-      return (int) Math.max(random.nextLong(), 0);
+      return random.nextInt();
     }
-    return (int) diff;
+    return diff;
```
Comment (gn-harness) [inline 400-rest/src/main/java/software/wings/service/impl/DelegateTaskServiceClassicImpl.java:299]: nit: just replace with d1.getValue() < d2.getValue()

File: 400-rest/src/main/java/software/wings/service/impl/DelegateTaskServiceClassicImpl.java [hunk 3]
```diff
@@ -798,10 +799,42 @@ protected List<String> getEligibleDelegateListOrderedNumberByTaskAssigned(
...
-    Map<String, Long> tasksCount =
-        delegateTasks.stream().collect(groupingBy(DelegateTask::getDelegateId, Collectors.counting()));
-    eligibleListOfDelegates.forEach(delegate -> tasksCount.computeIfAbsent(delegate, key -> 0L));
-    return tasksCount.entrySet().stream().sorted(delegateTaskCountComparator).map(Map.Entry::getKey).collect(toList());
+    Map<String, Integer> tasksCount =
+        delegateTasks.stream().collect(groupingBy(DelegateTask::getDelegateId, summingInt(delegateTaskCount -> 1)));
+    eligibleListOfDelegates.forEach(delegate -> tasksCount.computeIfAbsent(delegate, key -> 0));
+    // find entry with minimum count and used that to split into 2 maps
+    Optional<Map.Entry<String, Integer>> minCount = tasksCount.entrySet().stream().min(delegateTaskCountComparator);
+    int minVal = minCount.isPresent() ? minCount.get().getValue() : 0;
```
Comment (gn-harness) [inline 400-rest/src/main/java/software/wings/service/impl/DelegateTaskServiceClassicImpl.java:807]: nit: Return early if eligibleListOfDelegates is empty, returning Empty list.

File: 400-rest/src/test/java/software/wings/service/impl/DelegateTaskServiceClassicImplTest.java [hunk 4]
```diff
@@ -72,6 +74,36 @@ public void sortedEligibleListTest() throws ExecutionException {
     assertThat(sortedList.get(3)).isEqualTo(delegate3.getUuid());
   }
```
Comment (gn-harness) [inline 400-rest/src/test/java/software/wings/service/impl/DelegateTaskServiceClassicImplTest.java:75]: Please add the following tests:

1. Total three delegates. One delegate with 1 task, and others with >1. Check the first one is the delegate with 1 task.
2. 4 delegates. Each with different number of task. Check if the returned list is sorted.
2. 9 delegates. One with 2 tasks, other two with 3 and 4 tasks each. Check 0th index is 2-task delegate, 2nd index delegate has 3 tasks and 5th index has a delegate with 4 tasks

**Review by gn-harness — COMMENTED** [review]
Looks good, please take of minor comments and add a few tests

File: 400-rest/src/test/java/software/wings/service/impl/DelegateTaskServiceClassicImplTest.java (hunk 1 above)
Comment (jennyjames21) [reply 400-rest/src/test/java/software/wings/service/impl/DelegateTaskServiceClassicImplTest.java:86]: new test added

File: 400-rest/src/main/java/software/wings/service/impl/DelegateTaskServiceClassicImpl.java (hunk 2 above)
Comment (jennyjames21) [reply 400-rest/src/main/java/software/wings/service/impl/DelegateTaskServiceClassicImpl.java]: updated

File: 400-rest/src/test/java/software/wings/service/impl/DelegateTaskServiceClassicImplTest.java (hunk 4 above)
Comment (jennyjames21) [reply 400-rest/src/test/java/software/wings/service/impl/DelegateTaskServiceClassicImplTest.java:76]: updated

**Review by gn-harness — APPROVED** [review]

**Review by gn-harness — DISMISSED** [review]

File: 400-rest/src/test/java/software/wings/service/impl/DelegateTaskServiceClassicImplTest.java [hunk 5]
```diff
@@ -104,6 +105,119 @@ public void sortedEligibleListTestRand() throws ExecutionException {
...
+            .status(DelegateTask.Status.QUEUED)
+            .stageId(generateUuid())
+            .taskDataV2(TaskDataV2.builder().taskType(TaskType.INITIALIZATION_PHASE.name()).build())
+            .build();
+    persistence.save(delegateTask);
+
+    List<String> sortedList = delegateTaskServiceClassic.getEligibleDelegateListOrderedNumberByTaskAssigned(
+        delegateTask,
+        Arrays.asList(delegate1.getUuid(), delegate2.getUuid(), delegate3.getUuid(), delegate4.getUuid(),
+            delegate5.getUuid(), delegate6.getUuid(), delegate7.getUuid(), delegate8.getUuid(), delegate9.getUuid()));
```
[the remark this answers is not in the document]
Comment (jennyjames21) [reply 400-rest/src/test/java/software/wings/service/impl/DelegateTaskServiceClassicImplTest.java]: updated

**Comment by jennyjames21** [comment]
trigger uts

**Review by gn-harness — APPROVED** [review]

**Pull request merged**